AwsCryptographicMaterialProviders/runtimes/java/src/testExamples/java/software/amazon/cryptography/example/hierarchy/concurrent/ConcurrentConditionCheckWriteTest.java

```diff
@@ -0,0 +1,156 @@
+      )
+    );
+  }
+
+  public static Map<String, AttributeValue> indexItem(
+    final AttributeValue value,
+    final String timestamp
+  ) {
+    Map<String, AttributeValue> item = new HashMap<>();
+
+    item.put("branch-key-id", AttributeValue.builder().s(mLockedId).build());
+    item.put("type", AttributeValue.builder().s(indexType()).build());
+    item.put("value", value);
+    item.put("timestamp", AttributeValue.builder().s(timestamp).build());
+    return item;
+  }
+
+  private static String indexType() {
+    return "branch:ACTIVE";
+  }
+
+  public static TransactWriteItem conditionalWrite(
+    final AttributeValue value,
+    final String timestamp
+  ) {
+    return TransactWriteItem
+      .builder()
+      .put(putBuilder ->
+        putBuilder
+          .tableName(Fixtures.TEST_KEYSTORE_NAME)
+          .item(indexItem(value, timestamp))
+          .conditionExpression("attribute_not_exists(#pk)")
+          .expressionAttributeNames(INDEX_EXPR_ATT_NAMES)
+      )
+      .build();
+  }
+
+  private DynamoDbClient clientForThread(final String threadIdToIndex) {
+    return threadIdToDdbClient.computeIfAbsent(
+      threadIdToIndex,
+      ddbClient -> DynamoDbClient.create()
+    );
+  }
+
+  @Test(threadPoolSize = 5, invocationCount = 30, timeOut = 1000)
+  public void TestConcurrentWriteCheck() {
+    String threadId = String.valueOf(Thread.currentThread().getId());
+    String threadIdToIndex = indexToThreadId.computeIfAbsent(
+      threadId,
+      str -> unpickedIndices.pop()
+    );
+    AttributeValue value = AttributeValue.builder().s(threadIdToIndex).build();
+    TimeZone tz = TimeZone.getTimeZone("UTC");
+    DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSSS'Z'"); // Quoted "Z" to indicate UTC, no timezone offset
+    df.setTimeZone(tz);
+    String timestamp = df.format(new Date());
+
+    System.out.println(
+      "Thread ID: " +
+      Thread.currentThread().getId() +
+      " ThreadIndex: " +
+      threadIdToIndex +
+      " Timestamp: " +
+      timestamp
+    );
+
+    try {
+      DynamoDbClient client = clientForThread(threadIdToIndex);
+      TransactWriteItemsResponse transactWriteItemsResponse =
+        client.transactWriteItems(
+          TransactWriteItemsRequest
+            .builder()
+            .transactItems(conditionalWrite(value, timestamp))
+            .build()
+        );
+      Assert.assertTrue(
+        transactWriteItemsResponse.sdkHttpResponse().isSuccessful()
+      );
+    } catch (TransactionCanceledException exception) {
+      // We can fail for two reasons, either there's already a transact write in flight
+      // 0r we have failed the condition check.
+      exception
+        .cancellationReasons()
+        .forEach(cancellationReason -> {
+          Assert.assertTrue(
+            (cancellationReason.code().equals("TransactionConflict") ||
+              cancellationReason.code().equals("ConditionalCheckFailed"))
+          );
+        });
+    }
```

Comment:
Blocking/Question: I am confused.
This test can pass under three scenarios:
- The Write was successful 
- The Write failed with TransactionConcflict
- The Write failed with ConditionalCheckFailed

But there is no guarantee that each of the instances will be thrown at least once during the 5 * 30 = 150 iterations.

Is that an issue?

Reply:
No. After spending hours looking at the logs and adjusting the amount of threads that run this test, those exceptions will be thrown. There are enough threads trying to write, only one will win. Each thread runs the test x times. It is very likely that by the 2 or 3 time it tries to write the thread will submit a request that there is either a transactionconflict or that an item was already written and we now fail because the conditional check failed. Any other error we treat as a legit failure.

Reply:
OK. It would not be hard to add to some more concurrent data structures to track the various test outcomes,
and then have a follow up test assert minimum populations of the various test outcomes. 

Which would make these tests valid without manual logging parsing. 
But we can leave that as a TODO for another time.